### ISSUE TITLE
Refactor _MODE_CONV into ImageModes

### ISSUE THREAD

**Issue by FirefoxMetzger** [python-pillow/Pillow#5990]
Currently, Image.py has a private variable called `_MODE_CONV` that looks like this:

https://github.com/python-pillow/Pillow/blob/8e4a618b550b1a67e31c5362552db74b2cc72841/src/PIL/Image.py#L211-L239

I find this constant quite useful, and I was wondering if this info can be made part of `ImageMode` and exposed publically, so that I can (confidently) use it in downstream code. I think it would also make this codebase cleaner overall :D

For my use case, I am pre-allocating buffers into which I later want to decode image data, i.e., I have something similar to [`_conv_type_shape`](https://github.com/python-pillow/Pillow/blob/8e4a618b550b1a67e31c5362552db74b2cc72841/src/PIL/Image.py#L242) and would like to keep it in sync with the modes supported by pillow (without having to update manually).

**Comment by radarhere** [comment]
I realise this probably doesn't resolve your request, but here is a way to get that information using public APIs.

```python
from PIL import Image
import numpy

for mode in ("1", "L", "LA", "I", "F", "P", "RGB", "RGBX", "RGBA", "CMYK", "YCbCr", "LAB", "HSV", "I;16"):
    typ_extra = Image._MODE_CONV[mode]

    a = numpy.array(Image.new(mode, (0, 0)))
    typ_extra_from_public_api = (a.dtype.str, None if a.ndim < 3 else a.shape[2])

    assert typ_extra_from_public_api == typ_extra
```

**Comment by FirefoxMetzger** [comment]
That's a very cool solution to the problem indeed. It solves my use-case, so this would be a workable solution from my side. 

That said, I do wonder though if the suggested refactor would still be worthwhile.

**Comment by FirefoxMetzger** [comment]
@radarhere I just tested the snipped you shared, and it fails for me when calling `np.array` on the image. I tested with pillow 8.4,0, pillow9, numpy 1.21.4 and numpy 1.22.1. Could you let me know which versions you are using when running this snippet?

**Comment by radarhere** [comment]
Oh, my mistake, it would have been because I was using main.
To use Pillow 9.0, just change the size of the image from (0, 0) to (1, 1).

**Comment by radarhere** [comment]
Thinking about this, rather than making `_MODE_CONV` public, it might be better to make `_conv_type_shape` public, a method rather than a dictionary, so that we have more control. That could be useful in the future if we ever need to deprecate anything.

`_conv_type_shape` takes an image as the input, rather than a mode. That seems right, because I think the approach of our API is generally about dealing with images, not modes.

But once I talk about a public method that returns the shape and type of a NumPy array from an image, I start to think it sounds the same as `__array__` and the code I shared earlier -
```python
a = numpy.array(Image.new(mode, (0, 0)))
typ_extra_from_public_api = (a.dtype.str, None if a.ndim < 3 else a.shape[2])
```

So I don't think I'm convinced about this one.

**Comment by FirefoxMetzger** [comment]
Not exposing a dict does indeed sound like the right thing to do. I was thinking along the lines of making `_Mode_CONV` into  a property of [`ModeDescriptor`](https://github.com/python-pillow/Pillow/blob/7b3f28047fe008d70bda1938e3178da388ecaa64/src/PIL/ImageMode.py#L20) instead of exposing the dict directly, e.g. `ImageMode.getmode("RGB").dtype` or whichever name seems natural. This would pull a private dict that - essentially - contains mode metadata out of `Image` and make it part of `ImageMode` (where I would have expected mode-related code to live).

For downstream code there is one distinct difference between returning shape and dtype information vs `__array__` and that is decoding of pixel data. This is costly for large images, and ultimately how I ended up filing this issue. This does get solved by your suggestion of using a dummy array though, so at the moment I am doing something like:

```python
if self._image.format == "GIF":
    # GIF mode is inconsistent in pillow9; P first then RGB or RGBA
    # use pallete mode instead
    mode = self._image.palette.mode
else:
    mode = self._image.mode

dummy = np.asarray(Image.new(mode, (1, 1)))

shape = list(dummy.shape)
shape[:2] = self._image.size[::-1]
shape = (self._image.n_frames, *shape) if index is None else tuple(shape)

return ImageProperties(
            shape=shape,
            dtype=dummy.dtype,
            ...
        )
```

I like to believe that this is already fairly readable for others, although this might be even more so:

```python
if self._image.format == "GIF":
    # GIF mode is inconsistent in pillow9; P first then RGB or RGBA
    # use pallete mode instead
    mode_info = ImageMode.getmode(self._image.palette.mode)
else:
    mode_info = ImageMode.getmode(self._image.mode)

if index is None:
    shape = (self._image.n_frames, self._image.height, self._image.width, len(mode.bands))
else:
    shape = (self._image.height, self._image.width, len(mode.bands))

return ImageProperties(
            shape=shape,
            dtype=mode_info.dtype,
            ...
        )
```

However, since this is mostly stylistic I can understand if you decide that it is not worth the effort. In that case, please feel free to close this issue so that we can stay organized :)

**Comment by radarhere** [comment]
I hadn't thought of ImageMode.

Nor had I pieced together that `extra` was `len(mode.bands)`. That means that we could clean up the code by no longer listing that value for each mode.

And at that point, if we made the typestr available to ImageMode, `_MODE_CONV` could be removed altogether.

That actually sounds like a neater arrangement for this information within Pillow. Thanks! I've created PR #6057.
